In [87]:
# Pre Req Imports
import pandas as pd
import io
import os
import csv
import numpy as np

data_path = os.path.join("Resources")

# Define Excel Sheets

In [12]:
# Crime
# Table 3 is the largest and has the most useful data, crimes by type per suburb/town/postcode for each year from 2013 to 2022
xlsx = pd.ExcelFile("Data_Tables_LGA_Recorded_Offences_Year_Ending_December_2022.xlsx")
# Crime_By_Police_Region_Pop = pd.read_excel(xlsx, "Table 01") # Offences recorded and rate per 100,000 population by police region and local government area - January 2013 to December 2022
# Crime_By_Offence_Type_Pop = pd.read_excel(xlsx, "Table 02") # Offences recorded and rate per 100,000 population by offence type, local government area and police service area - January 2013 to December 2022
Crime_By_Offence_Type = pd.read_excel(xlsx, "Table 03") # Offences recorded by offence type, local government area and postcode or suburb/town - January 2013 to December 2022
# Crime_By_Location_Area = pd.read_excel(xlsx, "Table 04") # Offences recorded by location type and local government area - January 2013 to December 2022
# Crime_By_Investigation_Status = pd.read_excel(xlsx, "Table 05") # Offences recorded by investigation status and local government area - January 2013 to December 2022
# Crime_By_Drugs = pd.read_excel(xlsx, "Table 06") # Select drug offences by drug type and local government area - January 2013 to December 2022

In [16]:
# Property
xls = pd.ExcelFile("CleanSuburb_HouseV2021A.xls")
Property_Data = pd.read_excel(xls)

In [17]:
#View the property dataframe. Can drop 'Median 2011' and 'Median 2012' as no crime data for these years. Can 'Change' and 'Growth' columns be dropped as not relevant to study questions?
Property_Data.head()

locality Median 2011 Median 2012 Median 2013 Median 2014 Median 2015  \
0    ABBOTSFORD      730000      714000      792500      862500      925000   
1    ABERFELDIE      994000      852500      947500     1045000     1207500   
2       AINTREE           -           -           -           -      600000   
3  AIREYS INLET      680000      634000      664000      625500      680000   
4  AIRPORT WEST      557500      495000      532000      575000      635000   

  Median 2016 Median 2017 Median 2018 Median 2019 Median 2020  Median 2021  \
0     1187500     1280000     1192500     1050000     1200000      1375000   
1     1300000     1471000     1498500     1390000     1520000      1880500   
2      600000      571000      557500      575000      640000       719000   
3      715000      737500      869000      985000     1132500      1775000   
4      742000      845000      845000      795000      812500       960000   

   Median 2022  Change 2020-2021  Change 2011-2021  Growth PA  
0    1517500.0              15.0              88.0        6.5  
1    1575000.0              24.0              89.0        6.6  
2     760000.0              12.0               NaN        NaN  
3    1750000.0              57.0             161.0       10.1  
4     919500.0              18.0              72.0        5.6

In [18]:
# Show the datatypes for the property dataframe. Most medians are 'Object' dtype. Missing values represented by a hyphen. These need cleaning and converting to 'float'.
Property_Data.dtypes

locality             object
Median 2011          object
Median 2012          object
Median 2013          object
Median 2014          object
Median 2015          object
Median 2016          object
Median 2017          object
Median 2018          object
Median 2019          object
Median 2020          object
Median 2021           int64
Median 2022         float64
Change 2020-2021    float64
Change 2011-2021    float64
Growth PA           float64
dtype: object

In [19]:
# Count elements in each column of property data. Does not show missing values as they are represented by '-' 
Property_Data.count()

locality            820
Median 2011         820
Median 2012         820
Median 2013         820
Median 2014         820
Median 2015         820
Median 2016         820
Median 2017         820
Median 2018         820
Median 2019         820
Median 2020         820
Median 2021         820
Median 2022         799
Change 2020-2021    819
Change 2011-2021    809
Growth PA           803
dtype: int64

In [3]:
# View the crimes dataframe 
Crime_By_Offence_Type.head()

Year Year ending Local Government Area  Postcode Suburb/Town Name  \
0  2022    December                Alpine      3691         Dederang   
1  2022    December                Alpine      3691         Dederang   
2  2022    December                Alpine      3691         Dederang   
3  2022    December                Alpine      3691         Dederang   
4  2022    December                Alpine      3691         Dederang   

                       Offence Division                   Offence Subdivision  \
0           A Crimes against the person       Other crimes against the person   
1     B Property and deception offences                             B40 Theft   
2     B Property and deception offences                             B40 Theft   
3  D Public order and security offences   D10 Weapons and explosives offences   
4  D Public order and security offences  D20 Disorderly and offensive conduct   

                     Offence Subgroup  Offence Count  
0     Other crimes against the person              2  
1      B42 Steal from a motor vehicle              1  
2                     B49 Other theft              1  
3               D11 Firearms offences              1  
4  D22 Drunk and disorderly in public              1

In [4]:
# Count elements in each column of the crimes dataframe

Year                     357852
Year ending              357852
Local Government Area    357852
Postcode                 357852
Suburb/Town Name         357852
Offence Division         357852
Offence Subdivision      357852
Offence Subgroup         357852
Offence Count            357852
dtype: int64

In [11]:
# Show the datatypes for columns in the crimes dataframe
Crime_By_Offence_Type.dtypes

Year                      int64
Year ending              object
Local Government Area    object
Postcode                  int64
Suburb/Town Name         object
Offence Division         object
Offence Subdivision      object
Offence Subgroup         object
Offence Count             int64
dtype: object

In [33]:
# Obtain count of unique values in each column and save in Resources folder.
# Can drop Year Ending column.
# 'Postcode' and 'Suburb/Town Name' do not have a one to one correspondence. 'Postcode' has a better numerical correspondence to 'locality' in Property dataframe
with open ("Resources/crime_unique_values.csv", "w", newline='') as f:
    writer = csv.writer(f)
    for col in Crime_By_Offence_Type.columns:
        crime_unique_values = Crime_By_Offence_Type[col].nunique()
        writer.writerow([col, crime_unique_values])
        print(col, crime_unique_values)

Year 10
Year ending 1
Local Government Area 79
Postcode 693
Suburb/Town Name 2851
Offence Division 6
Offence Subdivision 25
Offence Subgroup 106
Offence Count 739


In [66]:
# Rank 'Offence Division' from most to least and output to Resources folder
offence_divs = Crime_By_Offence_Type.groupby("Offence Division")["Offence Count"].sum().reset_index()
offence_divs.to_csv("Resources/offence_divs.csv", index=False)
offence_divs

Offence Division  Offence Count
0           A Crimes against the person         769284
1     B Property and deception offences        2819417
2                       C Drug offences         304883
3  D Public order and security offences         339277
4         E Justice procedures offences         688660
5                      F Other offences          62866

In [8]:
# Rank 'Offence Subgroup' from most to least and output to Resources folder
offence_subgroups = Crime_By_Offence_Type.groupby("Offence Subgroup")["Offence Count"].sum().reset_index().sort_values("Offence Count", ascending=False)
offence_subgroups.to_csv("Resources/offence_subgroups.csv", index=False)
offence_subgroups.head()

Offence Subgroup  Offence Count
30    B42 Steal from a motor vehicle         536486
35                   B49 Other theft         457996
78  E21 Breach family violence order         391504
20               B21 Criminal damage         376218
38   B53 Obtain benefit by deception         290343

In [9]:
# Rank 'Offence Subdivision' from most to least and output to Resources folder
offence_subdivs = Crime_By_Offence_Type.groupby("Offence Subdivision")["Offence Count"].sum().reset_index().sort_values("Offence Count", ascending=False)
offence_subdivs.to_csv("Resources/offence_subdivs.csv", index=False)
offence_subdivs.head()

Offence Subdivision  Offence Count
7                          B40 Theft        1613247
19            E20 Breaches of orders         578829
6       B30 Burglary/Break and enter         421525
0   A20 Assault and related offences         420295
5                B20 Property damage         412405

In [10]:
# Rank 'Offence Division', 'Subdivision' and 'Subgroup' from most to least and output to Resources folder
offence_divs_subdivs_subgroup = Crime_By_Offence_Type.groupby(["Offence Division","Offence Subdivision","Offence Subgroup"])["Offence Count"].sum().reset_index().sort_values(by=["Offence Division", "Offence Count"], ascending=[True, False])
offence_divs_subdivs_subgroup.to_csv("Resources/offence_divs_subdivs_subgroup.csv", index=False)
offence_divs_subdivs_subgroup.head()

Offence Division               Offence Subdivision  \
17  A Crimes against the person   Other crimes against the person   
3   A Crimes against the person  A20 Assault and related offences   
4   A Crimes against the person  A20 Assault and related offences   
1   A Crimes against the person  A20 Assault and related offences   
0   A Crimes against the person  A20 Assault and related offences   

                   Offence Subgroup  Offence Count  
17  Other crimes against the person         141823  
3            A231 FV Common assault         135573  
4        A232 Non-FV Common assault          96442  
1       A212 Non-FV Serious assault          81752  
0           A211 FV Serious assault          77523

# Summary of table reviews
## Property


## Crime
The best crimes category to use for breaking the crimes data apart after merging with median house price data would be the highest level ***Offense Divisions*** column, consisting of the following divisions with total offence counts over the full 10 year period. The largest subdivision within each division, and subgroup within each subdivision is also shown, with respective offence count.
| Offense Division | Offence Count | Largest Offense Subdivision | Offence Count | Largest Offence Subgroup | Offence Count |
| ----------- | ----------- | ----------- | ----------- | ----------- | ----------- |
| A Crimes against the person | 769284 | A20 Assault and related offences | 420295 | A231 FV Common assault | 135573 |
| B Property and deception offences | 2819417 | B40 Theft | 1613247 | B42 Steal from a motor vehicle | 536486 |
| C Drug offences | 304883 | C30 Drug use and possession | 240559 | C32 Drug possession | 232063 |
| D Public order and security offences | 339277 | D20 Disorderly and offensive conduct | 151926 | D22 Drunk and disorderly in public | 72421 |
| E Justice procedures offences | 688660 | E20 Breaches of orders | 578829 | E13 Resist or hinder officer | 76228 |
| F Other offences | 62866 | F90 Miscellaneous offences | 52578 | F92 Public health and safety offences | 48932 |

We can probably omit ***F Other offences***.

In [37]:
# Convert 'locality' to title case in order to merge with 'Suberb/Town Name'
Property_Data['locality'] = Property_Data['locality'].astype(str).str.title()
Property_Data.head()

locality Median 2011 Median 2012 Median 2013 Median 2014 Median 2015  \
0    Abbotsford      730000      714000      792500      862500      925000   
1    Aberfeldie      994000      852500      947500     1045000     1207500   
2       Aintree           -           -           -           -      600000   
3  Aireys Inlet      680000      634000      664000      625500      680000   
4  Airport West      557500      495000      532000      575000      635000   

  Median 2016 Median 2017 Median 2018 Median 2019 Median 2020  Median 2021  \
0     1187500     1280000     1192500     1050000     1200000      1375000   
1     1300000     1471000     1498500     1390000     1520000      1880500   
2      600000      571000      557500      575000      640000       719000   
3      715000      737500      869000      985000     1132500      1775000   
4      742000      845000      845000      795000      812500       960000   

   Median 2022  Change 2020-2021  Change 2011-2021  Growth PA  
0    1517500.0              15.0              88.0        6.5  
1    1575000.0              24.0              89.0        6.6  
2     760000.0              12.0               NaN        NaN  
3    1750000.0              57.0             161.0       10.1  
4     919500.0              18.0              72.0        5.6

In [41]:
# Delete columns for 2011 and 2012 plus the Change and Growth columns
Property_Data.drop(['Median 2011', 'Median 2012', 'Change 2020-2021', 'Change 2011-2021', 'Growth PA'], axis = 1, inplace=True)
Property_Data.head()

locality Median 2013 Median 2014 Median 2015 Median 2016 Median 2017  \
0    Abbotsford      792500      862500      925000     1187500     1280000   
1    Aberfeldie      947500     1045000     1207500     1300000     1471000   
2       Aintree           -           -      600000      600000      571000   
3  Aireys Inlet      664000      625500      680000      715000      737500   
4  Airport West      532000      575000      635000      742000      845000   

  Median 2018 Median 2019 Median 2020  Median 2021  Median 2022  
0     1192500     1050000     1200000      1375000    1517500.0  
1     1498500     1390000     1520000      1880500    1575000.0  
2      557500      575000      640000       719000     760000.0  
3      869000      985000     1132500      1775000    1750000.0  
4      845000      795000      812500       960000     919500.0

In [46]:
# Convert all the Median house price columns to float datatype, or NaN if the value is a string
Property_Data[['Median 2013', 'Median 2014', 'Median 2015', 'Median 2016', 'Median 2017', 'Median 2018', 'Median 2019', 'Median 2020', 'Median 2021', 'Median 2022']] \
    = Property_Data[['Median 2013', 'Median 2014', 'Median 2015', 'Median 2016', 'Median 2017', 'Median 2018', 'Median 2019', 'Median 2020', 'Median 2021', 'Median 2022']] \
    .apply(pd.to_numeric, errors='coerce', downcast='float')
Property_Data.dtypes

locality        object
Median 2013    float32
Median 2014    float32
Median 2015    float32
Median 2016    float32
Median 2017    float32
Median 2018    float32
Median 2019    float32
Median 2020    float32
Median 2021    float32
Median 2022    float32
dtype: object

In [48]:
# Some rows contain null values but get rid of these after merging with crime dataframe
Property_Data.count()

locality       820
Median 2013    810
Median 2014    813
Median 2015    814
Median 2016    815
Median 2017    818
Median 2018    819
Median 2019    819
Median 2020    819
Median 2021    820
Median 2022    799
dtype: int64

In [49]:
# Delete the 'Year ending' column which just contains the value 'December' in every row
del Crime_By_Offence_Type['Year ending']
Crime_By_Offence_Type.head()

Year Local Government Area  Postcode Suburb/Town Name  \
0  2022                Alpine      3691         Dederang   
1  2022                Alpine      3691         Dederang   
2  2022                Alpine      3691         Dederang   
3  2022                Alpine      3691         Dederang   
4  2022                Alpine      3691         Dederang   

                       Offence Division                   Offence Subdivision  \
0           A Crimes against the person       Other crimes against the person   
1     B Property and deception offences                             B40 Theft   
2     B Property and deception offences                             B40 Theft   
3  D Public order and security offences   D10 Weapons and explosives offences   
4  D Public order and security offences  D20 Disorderly and offensive conduct   

                     Offence Subgroup  Offence Count  
0     Other crimes against the person              2  
1      B42 Steal from a motor vehicle              1  
2                     B49 Other theft              1  
3               D11 Firearms offences              1  
4  D22 Drunk and disorderly in public              1

In [75]:
# All 'localities' Merge crime and property using an inner join.
crime_property_merge = pd.merge(Crime_By_Offence_Type, Property_Data, left_on='Suburb/Town Name', right_on = 'locality', how = 'outer', indicator = True)
crime_property_merge.to_csv('Resources/merge_attempt_2.csv')
crime_property_merge.head()


Year Local Government Area  Postcode Suburb/Town Name  \
0  2022.0                Alpine    3691.0         Dederang   
1  2022.0                Alpine    3691.0         Dederang   
2  2022.0                Alpine    3691.0         Dederang   
3  2022.0                Alpine    3691.0         Dederang   
4  2022.0                Alpine    3691.0         Dederang   

                       Offence Division                   Offence Subdivision  \
0           A Crimes against the person       Other crimes against the person   
1     B Property and deception offences                             B40 Theft   
2     B Property and deception offences                             B40 Theft   
3  D Public order and security offences   D10 Weapons and explosives offences   
4  D Public order and security offences  D20 Disorderly and offensive conduct   

                     Offence Subgroup  Offence Count locality  Median 2013  \
0     Other crimes against the person            2.0      NaN          NaN   
1      B42 Steal from a motor vehicle            1.0      NaN          NaN   
2                     B49 Other theft            1.0      NaN          NaN   
3               D11 Firearms offences            1.0      NaN          NaN   
4  D22 Drunk and disorderly in public            1.0      NaN          NaN   

   Median 2014  Median 2015  Median 2016  Median 2017  Median 2018  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   

   Median 2019  Median 2020  Median 2021  Median 2022     _merge  
0          NaN          NaN          NaN          NaN  left_only  
1          NaN          NaN          NaN          NaN  left_only  
2          NaN          NaN          NaN          NaN  left_only  
3          NaN          NaN          NaN          NaN  left_only  
4          NaN          NaN          NaN          NaN  left_only

In [91]:
print("Test joining by OUTER join on Crime['Suburb/Town Name'] <-> Property['locality']")
print("--------------------------------------------------------------------------------")
print(f"Length of Crime table: {len(Crime_By_Offence_Type)}")
print(f"Length of Property table: {len(Property_Data)}")
print(f"Length of merged table: {len(crime_property_merge)}")
print(f"Merged from both tables: {(crime_property_merge._merge == 'both').sum()}")
print(f"Merged from Crime table only: {(crime_property_merge._merge == 'left_only').sum()}")
print(f"Merged from Property table only: {(crime_property_merge._merge == 'right_only').sum()}")
orphan_suburbs = crime_property_merge[crime_property_merge['_merge'] == 'left_only']
orphan_localities = crime_property_merge[crime_property_merge['_merge'] == 'right_only']
matched_suburbs = crime_property_merge[crime_property_merge['_merge'] == 'both']
matched_localities = crime_property_merge[crime_property_merge['_merge'] == 'both']
print(f"{orphan_suburbs['Suburb/Town Name'].nunique()} of {Crime_By_Offence_Type['Suburb/Town Name'].nunique()} unique suburbs in Crime table have no property data")
print(f"{orphan_localities['locality'].nunique()} of {Property_Data['locality'].nunique()} unique localities in Property table have no crime data")
print(f"{matched_suburbs['Suburb/Town Name'].nunique()} of {Crime_By_Offence_Type['Suburb/Town Name'].nunique()} unique suburbs in Crime table match localities in Property table")
print(f"{matched_localities['locality'].nunique()} of {Property_Data['locality'].nunique()} unique localities in Property table match suburbs in Crime table")

Test joining by OUTER join on Crime['Suburb/Town Name'] <-> Property['locality']
--------------------------------------------------------------------------------
Length of Crime table: 357852
Length of Property table: 820
Length of merged table: 357894
Merged from both tables: 273479
Merged from Crime table only: 84373
Merged from Property table only: 42
2073 of 2851 unique suburbs in Crime table have no property data
42 of 820 unique localities in Property table have no crime data
778 of 2851 unique suburbs in Crime table match localities in Property table
778 of 820 unique localities in Property table match suburbs in Crime table


In [90]:
# Save the unmatched localites and suburbs to csv files
unique_orphan_suburbs = orphan_suburbs['Suburb/Town Name'].unique()
pd.DataFrame(unique_orphan_suburbs).to_csv('Resources/orphan_suburbs.csv')
unique_orphan_localities = orphan_localities['locality'].unique()
pd.DataFrame(unique_orphan_localities).to_csv('Resources/orphan_localities.csv')

In [64]:
# Create six different crime_property_merge csv files from the main merge file above, each one named for, and only containing rows corresponding to one of the six 'Offence Division's
offence_divs = crime_property_merge['Offence Division'].unique()
for divs in offence_divs:
    crime_divs_property_merge = crime_property_merge[crime_property_merge['Offence Division'] == divs]
    crime_divs_property_merge.to_csv(f'Resources/{divs}.csv')

# Merged Data
The `Resources` folder contains the following working csv files:
- `A Crimes against the person.csv`
- `B Property and deception offences.csv`
- `C Drug offences.csv`
- `D Public order and security offences.csv`
- `E Justice procedures offences.csv`
- `F Other offences.csv`

Each is a merge of cleaned crimes left joined to cleaned property. As such there are several rows where there is no propoerty data. These can easily be removed with dropna.

Other files in the Resources folder were created to understand the structure of the raw source files and are not so important.

